server/src/main/java/com/cloud/ha/HighAvailabilityManagerImpl.java

```diff
@@ -680,31 +685,51 @@ public void cancelDestroy(VMInstanceVO vm, Long hostId) {
         _haDao.delete(vm.getId(), WorkType.Destroy);
     }
 
+    private void stopVMWithCleanup(VirtualMachine vm, VirtualMachine.State state) throws OperationTimedoutException, ResourceUnavailableException {
+        if (VirtualMachine.State.Running.equals(state)) {
+            _itMgr.advanceStop(vm.getUuid(), true);
+        }
+    }
+
+    private void destroyVM(VirtualMachine vm, boolean expunge) throws OperationTimedoutException, AgentUnavailableException {
+        s_logger.info("Destroying " + vm.toString());
+        if (VirtualMachine.Type.ConsoleProxy.equals(vm.getType())) {
+            consoleProxyManager.destroyProxy(vm.getId());
+        } else if (VirtualMachine.Type.SecondaryStorageVm.equals(vm.getType())) {
+            secondaryStorageVmManager.destroySecStorageVm(vm.getId());
+        } else {
+            _itMgr.destroy(vm.getUuid(), expunge);
+        }
+    }
+
     protected Long destroyVM(final HaWorkVO work) {
         final VirtualMachine vm = _itMgr.findById(work.getInstanceId());
-        s_logger.info("Destroying " + vm.toString());
+        if (vm == null) {
+            s_logger.info("No longer can find VM " + work.getInstanceId() + ". Throwing away " + work);
+            return null;
+        }
+        boolean expunge = VirtualMachine.Type.SecondaryStorageVm.equals(vm.getType())
```

Comment:
```suggestion
        boolean expungeSystemVM = VirtualMachine.Type.SecondaryStorageVm.equals(vm.getType())
```

Reply:
@sureshanaparti I think `expunge` name is better here as it is use irrespective of vm type in `destroyVM` method. Through it holds value - `true` only for ssvm and cpvm